### PR TITLE
Create custom user model and define roles

### LINKED ISSUE #63: Add User Model
## Overview

- Should be a custom model that extends default Django `auth.user`
- Should have a field for user role, which should take one of three values: `contributor`, `validator`, `administrator`

A custom user model often needs to be the very first migration in Django by design: https://docs.djangoproject.com/en/dev/topics/auth/customizing/#extending-the-existing-user-model, so we should do this first.

Consider if roles should be enums or a redefinable entity in the database.

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Overview
Create a custom user model, and provide a data migration to supply initial roles (CONTRIBUTOR, VALIDATOR, and ADMINISTRATOR).

By including a custom user model in the initial migration, we preserve the ability to customize the user model in the future.

Closes #63

### Demo
edit-user-in-admin

### Notes

- This custom user inherits from `AbstractBaseUser`, and reimplements most of the functionality on `AbstractUser` but for anything involving usernames. We're using email as the unique identifier rather than asking users to supply a username.
- Django reference for customizing a user model: https://docs.djangoproject.com/en/3.2/topics/auth/customizing/#specifying-custom-user-model
- I defined roles with a model, and I provided an initial data migration for `ADMINISTRATOR`, `VALIDATOR`, and `CONTRIBUTOR`.
- In `Role.clean()` I upper-case the roles to guard against mixed-case role definitions in the future (although unlikely!)
- I chose `CONTRIBUTOR` as the default Role to reduce clicks when using the Django admin to add users, see demo.

## Testing Instructions

- It may be necessary to unmigrate the `admin` app if you have already applied its migrations (do so with `scripts/manage migrate admin zero`)
- `scripts/manage migrate`
- `scripts/server`
- `scripts/manage createsuperuser` to create a superuser who can log in to the Django admin
- Visit Django admin page for User objects: http://localhost:8181/admin/api/user/
  - [x] Ensure your superuser exists and is editable
  - [x] Ensure generic users can be created
  - [x] Ensure "CONTRIBUTOR" is defaulted in the Role picker

## Checklist

- [ ] `fixup!` commits have been squashed
- [x] `CHANGELOG.md` updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines
- [ ] `README.md` updated if necessary to reflect the changes
- [ ] CI passes after rebase

### CODE REVIEW

File: src/django/api/models.py [hunk 1]
```diff
@@ -1,4 +1,78 @@
-# from django.db import models
```
[the remark this answers is not in the document]
Comment (mstone121) [reply src/django/api/models.py:1]: I guess there is a section in the documentation: https://docs.djangoproject.com/en/3.2/topics/db/models/#organizing-models-in-a-package

[the remark this answers is not in the document]
Comment (rajadain) [reply src/django/api/models.py:1]: There is precedence for using multiple apps at Azavea: https://github.com/WikiWatershed/model-my-watershed/tree/develop/src/mmw/apps.

I agree that having very long files can be unwieldy. Will suggest a way to do it in a way that doesn't require #69 to be done before this.

**Comment by rajadain** [comment]
I added a suggestion for how to split the models into their own files in [80dde91](https://github.com/azavea/iow-boundary-tool/pull/72/commits/80dde91fea18a2722885858e4f2adeef23ead33d). It was easiest to do so in a commit, but feel free to do make changes or offer improvements.

Also, can the default user role for a new user in the Django Admin interface be CONTRIBUTOR? As it is in the model?

**Comment by jacobtylerwalls** [comment]
> Also, can the default user role for a new user in the Django Admin interface be CONTRIBUTOR? As it is in the model?

Yes! That was an oversight in the testing instructions. I believe the default is as you describe: CONTRIBUTOR. (I'll check.)

**Comment by rajadain** [comment]
The default is ADMINISTRATOR on my end:

image

File: src/django/api/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jacobtylerwalls) [reply src/django/api/models.py:1]: Thanks for the link, Matt--that's a good way to organize models under a single app. I _think_ it's too early to know whether we will need multiple apps.

File: src/django/api/models/role.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+from enum import Enum, unique
+from django.db import models
+
+__all__ = ["Role"]
+
+
+@unique
+class Roles(Enum):
```
Comment (rajadain) [inline src/django/api/models/role.py:8]: 👏

**Review by rajadain — APPROVED** [review]
This looks great! I like the way you did it better.

It would be good to get Black or something like that going for this project to standardize quotes, small things like that. That can be done later.

**Pull request merged**